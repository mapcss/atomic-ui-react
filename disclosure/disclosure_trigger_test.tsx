import DisclosureTrigger from "./disclosure_trigger.ts";
import Disclosure from "./disclosure.ts";
import SSRProvider from "../ssr/ssr_provider.ts";
import {
  anyBoolean,
  anyFunction,
  anyString,
  assertSnapshot,
  describe,
  expect,
  fn,
  it,
  setupJSDOM,
} from "../dev_deps.ts";
import { fireEvent, render } from "@testing-library/react";

const describeTests = describe({
  name: "DisclosureTrigger",
  async beforeEach() {
    await setupJSDOM();
  },
});

it(describeTests, "should throw error when it not wrap Context", () => {
  expect(() => render(<DisclosureTrigger>test</DisclosureTrigger>)).toThrow();
});

it(
  describeTests,
  "should trigger click event by default",
  (t) => {
    const { baseElement, getByTestId } = render(
      <DisclosureTrigger data-testid="test">test</DisclosureTrigger>,
      {
        wrapper: ({ children }) => {
          return (
            <SSRProvider>
              <Disclosure>{children as never}</Disclosure>
            </SSRProvider>
          );
        },
      },
    );

    assertSnapshot(t, baseElement.innerHTML);

    const el = getByTestId("test");

    fireEvent.click(el);
    assertSnapshot(t, baseElement.innerHTML);

    fireEvent.click(el);
    assertSnapshot(t, baseElement.innerHTML);
  },
);

it(
  describeTests,
  "should change trigger event",
  (t) => {
    const { baseElement, getByTestId } = render(
      <DisclosureTrigger
        event={["mouseenter" as const, "mouseleave" as const]}
        data-testid="test"
      >
        test
      </DisclosureTrigger>,
      {
        wrapper: ({ children }) => {
          return (
            <SSRProvider>
              <Disclosure>{children as never}</Disclosure>
            </SSRProvider>
          );
        },
      },
    );

    const el = getByTestId("test");

    fireEvent.click(el);
    assertSnapshot(t, baseElement.innerHTML);

    fireEvent.mouseEnter(el);
    assertSnapshot(t, baseElement.innerHTML);

    fireEvent.mouseLeave(el);
    assertSnapshot(t, baseElement.innerHTML);
  },
);

it(
  describeTests,
  "should pass context as render children",
  () => {
    const mockFn = fn();
    render(
      <DisclosureTrigger>
        {(context) => {
          mockFn(context);
          return <></>;
        }}
      </DisclosureTrigger>,
      {
        wrapper: ({ children }) => {
          return (
            <SSRProvider>
              <Disclosure>{children as never}</Disclosure>
            </SSRProvider>
          );
        },
      },
    );

    expect(mockFn).toHaveBeenCalledWith({
      id: anyString(),
      isOpen: anyBoolean(),
      toggle: anyFunction(),
      open: anyFunction(),
      close: anyFunction(),
    });
  },
);

it(
  describeTests,
  "should render as children",
  (t) => {
    const { baseElement } = render(
      <DisclosureTrigger>
        {({ isOpen }) => {
          return (
            <span
              className={isOpen ? "i-mdi-chevron-up" : "i-mdi-chevron-down"}
            />
          );
        }}
      </DisclosureTrigger>,
      {
        wrapper: ({ children }) => {
          return (
            <SSRProvider>
              <Disclosure>{children as never}</Disclosure>
            </SSRProvider>
          );
        },
      },
    );
    assertSnapshot(t, baseElement.innerHTML);
  },
);
