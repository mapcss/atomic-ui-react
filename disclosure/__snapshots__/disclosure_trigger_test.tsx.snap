export const snapshot = {};

snapshot[`DisclosureTrigger > should trigger click event by default 1`] = `'<div><button aria-controls="atomic-ui-0" aria-expanded="false" data-testid="test">test</button></div>'`;

snapshot[`DisclosureTrigger > should trigger click event by default 2`] = `'<div><button aria-controls="atomic-ui-0" aria-expanded="true" data-testid="test">test</button></div>'`;

snapshot[`DisclosureTrigger > should trigger click event by default 3`] = `'<div><button aria-controls="atomic-ui-0" aria-expanded="false" data-testid="test">test</button></div>'`;

snapshot[`DisclosureTrigger > should change trigger event 1`] = `'<div><button aria-controls="atomic-ui-0" aria-expanded="false" data-testid="test">test</button></div>'`;

snapshot[`DisclosureTrigger > should change trigger event 2`] = `'<div><button aria-controls="atomic-ui-0" aria-expanded="true" data-testid="test">test</button></div>'`;

snapshot[`DisclosureTrigger > should change trigger event 3`] = `'<div><button aria-controls="atomic-ui-0" aria-expanded="false" data-testid="test">test</button></div>'`;

snapshot[`DisclosureTrigger > should render as children 1`] = `'<div><button aria-controls="atomic-ui-0" aria-expanded="false"><span class="i-mdi-chevron-down"></span></button></div>'`;
